Feature/ BJS2-41788 Notification Service: User Service Recommendation Event Publisher

drewbal opened the pull request:
New DTO for notifications;
New endpoint and service method;

Commit:
Feature / Notification Service: User Service Publisher

New DTO for notifications;
New endpoint and service method;

Commit:
Feature / Notification Service: User Service Publisher

Redis config

Commit:
Feature / Notification Service: User Service Publisher

adds Recommendation Received Event;
adds Publisher;
adds publish logic in create method in service class;

Darkselect commented on src/test/java/school/faang/user_service/controller/UserControllerTest.java:
```diff
@@ -26,39 +28,44 @@
     @Spy
     private UserMapper userMapper;
-    @Mock
+    @MockBean
```
В чем разница @Mock и @MockBean ?

drewbal replied:
разные вещи. @Mock - это чисто Mockito, изолированный юнит-тест, а @MockBean - это Спринговый бин, который мокает реальные Бины при поднятии контекста. Я подкрутил этот тест, чтобы сделать его настоящим MockMVC, поднимающим часть контекста. Без этого имеющаяся реализация не могла протестировать отрицательные сценарии, не видела вариаций и глобального обработчика

DeemaP replied:
Вот спасибо тебе, милый человек, я ночью тогда думал какого хрена отрицательные сценарии в тестах не получаются. :D
счастья тебе и здоровья.

Darkselect commented on src/main/java/school/faang/user_service/service/UserService.java:
```diff
@@ -54,6 +57,7 @@ public class UserService {
     public UserService(UserRepository userRepository,
```
А почему через @Autowired  ? а не @RequiredArgsConstructor

drewbal replied:
Не спрашивай :) тот же вопрос. В целом работает, но выглядит не лучшим образом. Я в этот класс добавил один метод, остальное не разбирал и не трогал. Тут надо хорошо рефакторить в будущем.

RuslanZinovyev replied:
Если мы делаем иньекцию зависимостей через конструктор то аннотацию RequredArgConstructor надо удалить.

drewbal replied:
Я сделал небольшой рефакторинг: убрал @Lazy зависимости и заменил на взаимодействие через @EventListener. 
Это позволило в том числе убрать конструктор.

Darkselect commented on src/main/java/school/faang/user_service/service/UserService.java:
```diff
@@ -54,6 +57,7 @@ public class UserService {
+                       UserContactsMapper userContactsMapper,
                        UserValidator userValidator,
                        CountryService countryService,
                        @Lazy MentorshipService mentorshipService,
```
@Lazy MentorshipService mentorshipService. Теперь понял почему

drewbal replied:
Но все равно тут можно подумать и сделать более изящно

RuslanZinovyev replied:
Тут можно подключить наше знание событийной архитектуры и избавиться от зависимости MemborshipService полностью.

drewbal replied:
Так и сделано. Теперь все красиво и по EDA

DeemaP commented on src/main/java/school/faang/user_service/dto/UserContactsDto.java:
```diff
@@ -0,0 +1,18 @@
+    private final String phone;
+    private final PreferredContact preference;
+
+    public enum PreferredContact {
```
Подскажи плиз, а зачем здесь эти enum если они у тебя в отдельном пакете как енам лежат? Тут хитрость какая-то?

drewbal replied:
А вот наверное и можно без него обойтись. Дело в том, что в NotificationService этот enum находится прям в дто и я просто скопипастил сюда, чтобы сохранить структуру

DeemaP commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -36,6 +39,13 @@ public RecommendationDto create(RecommendationDto recommendationDto) {
+                        recommendation.getAuthor().getId()
+                )
+        );
+
```
чего логирование по пути не прикрутил?)
Хотя тут его везде надо прикручивать, я понял....

drewbal replied:
Хорошо, что напомнил. Логи я нигде не сделал

RuslanZinovyev replied:
Так где логи в итоге? =)

drewbal replied:
Были написаны молоком :) Я подержал над свечой и все видно теперь :)

DeemaP reviewed (commented):
Отлично) Только вот я еще узнал о более современном аналоге Jedis - Lettuce, который якобы современный, поддерживает многопоточку и весь такой крутой. 
Не слышал о таком? Не нужна ли нам такая задачка со звездочкой? Прикрутить современный клиент для работы с редисом 😵‍💫

Commit:
Feature / Notification Service: User Service Publisher

Cross-review fixes

RuslanZinovyev commented on src/main/java/school/faang/user_service/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,37 @@
+    private final RedisProperties redisProperties;
+
+    @Bean
+    JedisConnectionFactory jedisConnectionFactory() {
```
Давай одной стратегии придерживаться, вот как бы имплементирован бин в сервисе уведомлений

```
 @Bean
    JedisConnectionFactory jedisConnectionFactory() {
        RedisStandaloneConfiguration configuration = new RedisStandaloneConfiguration();
        configuration.setHostName(redisProperties.getHost());
        configuration.setPort(redisProperties.getPort());
        return new JedisConnectionFactory(configuration);
    }
```

RuslanZinovyev replied:
Вообще тут у вас с Юрой конфликты будут ибо конфигурация должна быть одна.

drewbal replied:
По формату согласен, скорректировал.

Конфликта почти не было, потому что у Юры интересный конфиг, в котором я не вижу настроек для подключения к самому Редису :)
Пока не трогаю, но чувствую тут надо будет переработать

RuslanZinovyev replied:
Потому что о создании JedisConnectionFactory позаботится autoconfiguration

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -36,6 +39,13 @@ public RecommendationDto create(RecommendationDto recommendationDto) {
         saveNewSkillOffers(recommendation);
         skillService.addGuarantee(recommendation);
 
+        recommendationReceivedEventPublisher.publish(new RecommendationReceivedEvent(
```
Тут потенциальная dual write проблема. Такая проблема может возникать когда 2 независимых друг от друга сервиса делают операции, один сервис делает запись в базу, другой публикует сообщение в рамках одной транзакции, только транзакция не касается внешнего вызова. Может случиться так, что запись в базу будет сделана, а вот сообщение не будет опубликовано по какой-либо причине и как итог ты найдешьс свое приложение в полуторном состоянии. Ты владеешь английским поэтому рекомендую почитать вот эту статью:
https://www.confluent.io/blog/dual-write-problem/
Я не предлагаю сейчас условный Outbox имплементировать, но давай хотя бы retry механизм добавим, да это анти паттерн, так как не спасет нас от потери данных при полном краше приложения, но хотя бы существенно увеличит вероятность потери данных и как итог не полной потери консистентности.

drewbal replied:
Спасибо за рекомендацию. Я пока дочитал до этого коммента своим умом пришел к этой проблеме и нагуглил, что она фундаментальная в программировании. Outbox очень интересен, но требует времени и мозгов. 
Пока я сделал @Retryable, да, это костыль, но пока хоть что-то

Commit:
Feature / Notification Service: User Service Publisher

Review fixes:
Removes @Lazy dependencies in userService;
Adds EDA approach to resolve cyclical dependencies;
Adds config for @Retryable;
Adds logs;
Refactors RedisConfig;
Minor syntax fixes;

Commit:
Merge branch 'Dragon-master-stream7' into feature/BJS2-41788-Recomm-Event-Publisher

# Conflicts:
#	src/main/java/school/faang/user_service/config/redis/RedisConfig.java
#	src/main/java/school/faang/user_service/controller/UserController.java
#	src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java
#	src/main/java/school/faang/user_service/service/UserService.java
#	src/main/resources/application.yaml
#	src/test/java/school/faang/user_service/controller/UserControllerTest.java
#	src/test/java/school/faang/user_service/service/UserServiceTest.java

Commit:
Feature / Notification Service: User Service Publisher

Resolves minor conflicts

Commit:
Feature / Notification Service: User Service Publisher

Resolves minor conflicts

Commit:
Merge remote-tracking branch 'origin/feature/BJS2-41788-Recomm-Event-Publisher' into feature/BJS2-41788-Recomm-Event-Publisher

# Conflicts:
#	src/test/java/school/faang/user_service/service/UserServiceTest.java

RuslanZinovyev commented on src/main/java/school/faang/user_service/publisher/RecommendationReceivedEventPublisher.java:
```diff
@@ -0,0 +1,22 @@
+    private final RedisProperties redisProperties;
+
+    @Override
+    public void publish(RecommendationReceivedEvent event) {
```
Я не понял куда ты в итоге retry логику прикрутил

drewbal replied:
Хаха, а тут забыл. Я сделал retry при работе с двумя lazy зависимостями, которые были в userService. А для это Publisher забыл. Исправил

Commit:
Feature / Notification Service: User Service Publisher

Review fixes

Commit:
Feature / Notification Service: User Service Publisher

Minor fixes
